File: CustomCollectionViewLayoutDemo/ViewController.m
//
//  ViewController.m
//  CustomCollectionViewLayoutDemo
//
//  Created by taolv on 15/8/13.
//  Copyright (c) 2015年 taolv365.ios.recycle. All rights reserved.
//

#import "ViewController.h"
#import "DoubleGridLayout.h"
#import "NumCollectionViewCell.h"
#import "NumCollectionReusableView.h"
#import "SepicalHeadCollectionReusableView.h"

@interface ViewController ()<UICollectionViewDataSource, UICollectionViewDelegateDoubleGridLayout>

@property (nonatomic, strong) NSArray *collectionData;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    DoubleGridLayout *layout = [[DoubleGridLayout alloc] init];
    layout.rowHeight = 30;
    layout.columnWidth = 50;
    layout.rowNumber = 30;
    layout.columnNumber = 10;
    layout.leftHeaderWidth = 80;
    layout.topHeaderHeight = 50;
    layout.supplyHeaderHeight = 40;
    layout.showHorizonLine = YES;
    layout.showVerticalLine = YES;
    
    NSMutableArray *tmpArr = [NSMutableArray arrayWithCapacity:300];
    for (int i = 0; i != 10; ++i) {
        for (int j = 0; j != 30; ++j) {
            [tmpArr addObject:[NSNumber numberWithInt:i * 30 + j]];
        }
    }
    self.collectionData = tmpArr;
    
    UICollectionView *collectionView = [[UICollectionView alloc] initWithFrame:self.view.bounds collectionViewLayout:layout];
    collectionView.delegate = self;
    collectionView.dataSource = self;
    [collectionView registerClass:[NumCollectionViewCell class] forCellWithReuseIdentifier:@"MY_CELL"];
    [collectionView registerClass:[NumCollectionReusableView class] forSupplementaryViewOfKind:UICollectionElementKindTopHeader withReuseIdentifier:@"TOP_HEAD"];
    [collectionView registerClass:[NumCollectionReusableView class] forSupplementaryViewOfKind:UICollectionElementKindLeftHeader withReuseIdentifier:@"LEFT_HEAD"];
    [collectionView registerNib:[UINib nibWithNibName:@"SepicalHeadCollectionReusableView" bundle:nil] forSupplementaryViewOfKind:UICollectionElementKindSupplyHeader withReuseIdentifier:@"SUPPLY_HEAD"];
    collectionView.backgroundColor = [UIColor whiteColor];
    collectionView.directionalLockEnabled = YES;
    [self.view addSubview:collectionView];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (NSInteger)collectionView:(UICollectionView *)view numberOfItemsInSection:(NSInteger)section
{
    return self.collectionData.count;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)cv cellForItemAtIndexPath:(NSIndexPath *)indexPath;
{
    NumCollectionViewCell *cell = [cv dequeueReusableCellWithReuseIdentifier:@"MY_CELL" forIndexPath:indexPath];
    
    NSNumber *num = _collectionData[indexPath.item];
    
    cell.numLabel.text = num.stringValue;
    
    return cell;
}

- (UICollectionReusableView *)collectionView:(UICollectionView *)collectionView viewForSupplementaryElementOfKind:(NSString *)kind atIndexPath:(NSIndexPath *)indexPath
{
    if ([kind isEqualToString:UICollectionElementKindTopHeader]) {
        NumCollectionReusableView *view = [collectionView dequeueReusableSupplementaryViewOfKind:kind withReuseIdentifier:@"TOP_HEAD" forIndexPath:indexPath];
        view.numLabel.text = [NSString stringWithFormat:@"top %ld", (long)indexPath.item];
        view.backgroundColor = [UIColor redColor];
        return view;
    } else if ([kind isEqualToString:UICollectionElementKindLeftHeader]) {
        NumCollectionReusableView *view = [collectionView dequeueReusableSupplementaryViewOfKind:kind withReuseIdentifier:@"LEFT_HEAD" forIndexPath:indexPath];
        view.numLabel.text = [NSString stringWithFormat:@"left %ld", (long)indexPath.item];
        view.backgroundColor = [UIColor greenColor];
        return view;
    } else {
        SepicalHeadCollectionReusableView *view = [collectionView dequeueReusableSupplementaryViewOfKind:kind withReuseIdentifier:@"SUPPLY_HEAD" forIndexPath:indexPath];
        view.label.text = @"hello, world";
        return view;
    }
}

@end

File: FloatFlexHeaderLayout/ViewController.swift
//
//  ViewController.swift
//  FloatFlexHeaderLayout
//
//  Created by john on 16/8/18.
//  Copyright © 2016年 taolv365.ios.recycle. All rights reserved.
//

import UIKit

class ViewController: UIViewController {
    
    @IBOutlet weak var collectionView: UICollectionView!
    
    var collectionData: [String]!

    override func viewDidLoad() {
        super.viewDidLoad()
        
        let layout = collectionView.collectionViewLayout as! FloatFlexCollectionViewLayout
        layout.headerReferenceSize = CGSizeMake(100, 100)
        layout.footerReferenceSize = CGSizeMake(70, 70)
        layout.floatHeaderRegion = CGRect(x: 30, y: 30, width: 50, height: 50)
        
        collectionView.registerNib(UINib(nibName: "FixSupplementaryHeadCollectionReusableView", bundle: nil), forSupplementaryViewOfKind: UICollectionElementKindSectionHeader, withReuseIdentifier: "head")
        collectionView.registerNib(UINib(nibName: "FixSupplementaryHeadCollectionReusableView", bundle: nil), forSupplementaryViewOfKind: UICollectionElementKindSectionFooter, withReuseIdentifier: "foot")
        collectionView.registerNib(UINib(nibName: "SupplyHeadCollectionReusableView", bundle: nil), forSupplementaryViewOfKind: UICollectionElementKindSectionHeaderFloat, withReuseIdentifier: "supplementaryHead")
        collectionView.registerNib(UINib(nibName: "NormalCollectionViewCell", bundle: nil), forCellWithReuseIdentifier: "reuseIdentifier")
        collectionData = ["airport-icon.png", "beer-icon.png", "bicep-icon.png", "bike-icon.png", "birthday-icon.png", "bloody-mary-icon.png", "book-icon.png", "brunch-icon.png", "candy-cane-icon.png", "cat-icon.png", "chicken-leg-icon.png", "chinese-take-out-icon.png", "christmas-tree-icon.png", "clapper-icon.png", "coffee-icon.png", "coffee-2-icon.png", "cosmo-icon.png", "croissant-icon.png", "dog-icon.png", "dreidel-icon.png"]
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        
        let layout = collectionView.collectionViewLayout as! FloatFlexCollectionViewLayout
        
        layout.sectionInset = UIEdgeInsets(top: 15, left: 15, bottom: 5, right: 15)
        // 每行3个
        let width = floor((collectionView.frame.width - layout.sectionInset.left - layout.sectionInset.right - layout.minimumInteritemSpacing * 2) / 3)
        layout.itemSize = CGSizeMake(width, width)
        
        layout.invalidateLayout()
    }

}

extension ViewController: UICollectionViewDelegateFlowLayout, UICollectionViewDataSource
{
    func numberOfSectionsInCollectionView(collectionView: UICollectionView) -> Int {
        return 2
    }
    
    func collectionView(collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return collectionData.count
    }
    
    func collectionView(collectionView: UICollectionView, cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCellWithReuseIdentifier("reuseIdentifier", forIndexPath: indexPath) as! NormalCollectionViewCell
        
        cell.paint.image = UIImage(named: collectionData[indexPath.item])
        
        return cell
    }
    
    func collectionView(collectionView: UICollectionView, viewForSupplementaryElementOfKind kind: String, atIndexPath indexPath: NSIndexPath) -> UICollectionReusableView {
        if kind == UICollectionElementKindSectionHeaderFloat {
            return collectionView.dequeueReusableSupplementaryViewOfKind(kind, withReuseIdentifier: "supplementaryHead", forIndexPath: indexPath)
        } else if kind == UICollectionElementKindSectionHeader {
            let view = collectionView.dequeueReusableSupplementaryViewOfKind(kind, withReuseIdentifier: "head", forIndexPath: indexPath) as! FixSupplementaryHeadCollectionReusableView
            
            view.text.text = "fixed header \(indexPath.section)"
            
            return view
        } else {
            let view = collectionView.dequeueReusableSupplementaryViewOfKind(kind, withReuseIdentifier: "foot", forIndexPath: indexPath) as! FixSupplementaryHeadCollectionReusableView
            
            view.text.text = "fixed footer \(indexPath.section)"
            
            return view
        }
    }

}

